### PR TITLE
Fuse OverlayFs implementation

### PR DESCRIPTION
**Pull request by WeiZhang555**
These commits add first implementation of userspace OverlayFs, which aims to add writable upper layer for Nydus, and makes Nydus capable to get more control over user's file writing behavior.

The implementation refers a lot to kernel Overlay Fs, but there're still some differences due to limitation of Fuse
protocol. And the first implementation is not perfect but could be a start.

It was tested under `xfstest` testsuite, about 300+ test cases were tested and only 25 failed as designed or being limited by Fuse. I'll add external CI process later to make `xfstest` workable later.

Let me know your review opinions.

### CODE REVIEW

**Comment by WeiZhang555** [comment]
xfstests CI action costs about 39min,  one example log: https://github.com/WeiZhang555/fuse-backend-rs/actions/runs/6322500147/job/17168241410

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@
 **/Cargo.lock
 **/.vscode
 /.idea
+
```
Comment (jiangliu) [inline .gitignore:6]: Unrelated changes?

File: Makefile [hunk 2]
```diff
@@ -54,3 +54,4 @@ smoke-macos: check-macos
 
 docker-smoke:
 	docker run --env RUST_BACKTRACE=1 --rm --privileged --volume ${current_dir}:/fuse-rs rust:1.68 sh -c "rustup component add clippy rustfmt; cd /fuse-rs; make smoke-all"
+
```
Comment (jiangliu) [inline Makefile:57]: ditto

File: src/api/filesystem/mod.rs [hunk 3]
```diff
@@ -30,6 +30,11 @@ pub use async_io::{AsyncFileSystem, AsyncZeroCopyReader, AsyncZeroCopyWriter};
 mod sync_io;
 pub use sync_io::FileSystem;
 
+#[cfg(all(any(feature = "fusedev", feature = "virtiofs"), target_os = "linux"))]
```
Comment (jiangliu) [inline src/api/filesystem/mod.rs:33]: It would be preferrable to mount overlayfs inside guest, so is there any need to turn on overlayfs for virtiofs?

Comment (WeiZhang555) [reply src/api/filesystem/mod.rs:33]: The idea of Fuse OverlayFS is to take over Container RootFs totally through Nydus and Fuse-backend-rs, so I suppose mounting overlayfs on host is strong requirement.

Besides, the OverlayFs replies heavily on Passthroughfs currently, so I copied #[cfg] for 'passthroughfs' from `src/lib.rs`:

```
#[cfg(all(any(feature = "fusedev", feature = "virtiofs"), target_os = "linux"))]
pub mod passthrough;
```

File: src/api/filesystem/overlay.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        // Use temp value to avoid moved 'parent'.
+        let ino: u64 = inode.into();
+
+        // Get attributes of the directory.
+        let (st, _d) = self.getattr(ctx, ino.into(), None)?;
+        if !is_dir(st) {
+            return Err(Error::from_raw_os_error(libc::ENOTDIR));
+        }
+
+        // Return Result<(is_opaque, break)>, if 'break' is true, stop the whole function.
```
Comment (imeoer) [inline src/api/filesystem/overlay.rs:136]: It seems that `is_opaque` and `break` must be equal, so looks like `break` is unnecessary.

Comment (WeiZhang555) [reply src/api/filesystem/overlay.rs:136]: Yes, you're right. I'll optimize this part, thanks!

**Comment by WeiZhang555** [comment]
Add unionmount-testsuite as Overlay function test cases, this gave us more confidence on the functionality.

File: src/overlayfs/mod.rs [hunk 5]
```diff
@@ -1,2432 +1,2227 @@
...
+    pub fn first_layer_inode(&self) -> (Arc<BoxedLayer>, bool, u64) {
+        let all_inodes = self.real_inodes.lock().unwrap();
+        let first = all_inodes.first();
+        match first {
+            Some(v) => (v.layer.clone(), v.in_upper_layer, v.inode),
+            None => panic!("BUG: dangling OverlayInode"),
+        }
+    }
+
+    pub fn childrens(&self) -> HashMap<String, Arc<OverlayInode>> {
```
[the remark this answers is not in the document]
Comment (WeiZhang555) [reply src/overlayfs/mod.rs:797]: Sounds quite reasonable, let me do some modification.

**Comment by WeiZhang555** [comment]
The Integration into Nydus can be seen in Nydus [PR](https://github.com/dragonflyoss/nydus/pull/1544)

**Comment by WeiZhang555** [comment]
ping @hhhhsdxxxx @jiangliu @eryugey

**Comment by hhhhsdxxxx** [comment]
lgtm

**Review by hhhhsdxxxx — APPROVED** [review]

**Review by eryugey — APPROVED** [review]
